### ISSUE TITLE
paddlepaddle 1.5.0 cross_entropy对不齐 其他框架

### ISSUE THREAD

**Issue by KeyKy** [PaddlePaddle/Paddle#20994]
这是结论图：softmax+cross_entropy 对不上pytorch的nn.CrossEntropyLoss，而且有明显的特点就是大于64的loss都变为64；图中分别是cross_entropy 和 nn.CrossEntropyLoss的输出
![8741d8d974eb9eb0f28e64fc9](https://user-images.githubusercontent.com/2967075/68109394-6717da00-ff25-11e9-8c35-cfd5e7018966.png)

这是特征的输入100x2：
[score_out_tr_np.npy.zip](https://github.com/PaddlePaddle/Paddle/files/3803212/score_out_tr_np.npy.zip)

这是上面数据通过paddle1.5.0的softmax层之后的结果：
[softmax_out_tr_np.npy.zip](https://github.com/PaddlePaddle/Paddle/files/3803214/softmax_out_tr_np.npy.zip)

这是分类的真实标签100x1：
[label_tr_np.npy.zip](https://github.com/PaddlePaddle/Paddle/files/3803217/label_tr_np.npy.zip)

这是paddle1.5.0的cross_entropy结果：
[ce_tr_np.npy.zip](https://github.com/PaddlePaddle/Paddle/files/3803218/ce_tr_np.npy.zip)

这是pytorch 0.4.1的nn.CrossEntropyLoss(reduce=False)的结果
```
loss_fn = nn.CrossEntropyLoss(reduce=False)
tmp_np = loss_fn(torch.from_numpy(score_out_tr_np), torch.from_numpy(label_tr_np[:, 0]))
```
[torch_res.npy.zip](https://github.com/PaddlePaddle/Paddle/files/3803225/torch_res.npy.zip)

**Comment by Xreki** [comment]
1. loss里面出现的最大值64，是当一个数字很小时，std::log计算得到的值是-64。可以做如下测试：
```cpp
#include <iostream>
#include <cmath>

int main() {
  float number_fp = 1.60381e-28;
  std::cout << "std::log=" << std::log(number_fp) << std::endl;
  double number_dp = 1.60381e-28;
  std::cout << "std::log=" << std::log(number_dp) << std::endl;
  return 0;
}
```

会发现两个输出值都是-64。

2. PyTorch的CrossEntropyLoss计算的并不是单纯的cross_entropy，而是softmax+cross_entropy，具体见文档：https://pytorch.org/docs/stable/nn.html?highlight=crossentropyloss#torch.nn.CrossEntropyLoss。softmax计算中对一个数取exp，cross_entropy中再取log，会产生一定的精度损失。Paddle有对应的API softmax_with_cross_entropy，具体见文档：https://www.paddlepaddle.org.cn/documentation/docs/zh/api_cn/layers_cn/softmax_with_cross_entropy_cn.html#softmax-with-cross-entropy

3. Paddle的softmax_with_cross_entropy实现中：CPU版本是先计算了softmax再计算cross_entropy，因此计算结果仍会有精度损失；GPU版本将两个算子的计算逻辑融合了，计算结果与PyTorch一致。。验证代码如下：
```python
def softmax_with_cross_entropy():
    score = fluid.layers.data(name='score', shape=[100, 2], dtype='float32', append_batch_size=False)
    label = fluid.layers.data(name='label', shape=[100, 1], dtype='int64', append_batch_size=False)
    loss = fluid.layers.softmax_with_cross_entropy(logits=score, label=label)

    score_np = np.load("score_out_tr_np.npy")
    label_np = np.load("label_tr_np.npy")

    place = fluid.CUDAPlace(0)
    exe = fluid.Executor(place)
    fetchs = exe.run(fluid.default_main_program(),
                     feed={"score": score_np, "label": label_np},
                     fetch_list=[loss])
    print(fetchs[0])
```

输出结果：
```
W1105 03:47:10.707700  2581 device_context.cc:235] Please NOTE: device: 0, CUDA Capability: 70, Driver API Version: 10.1, Runtime API Version: 9.0
W1105 03:47:10.712313  2581 device_context.cc:243] device: 0, cuDNN Version: 7.5.
[[-0.      ]
 [-0.      ]
 [53.57935 ]
 [25.454594]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [35.79542 ]
 [20.844778]
 [-0.      ]
 [46.858597]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [33.83307 ]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [37.690758]
 [-0.      ]
 [41.716484]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [83.61784 ]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [72.865234]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [22.132366]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [36.45475 ]
 [61.83087 ]
 [38.043434]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [67.81977 ]
 [-0.      ]
 [-0.      ]
 [51.52208 ]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [35.1622  ]
 [27.061619]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [31.814034]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [43.25881 ]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [23.904854]
 [-0.      ]
 [56.074005]
 [-0.      ]
 [-0.      ]
 [41.209007]
 [-0.      ]
 [-0.      ]
 [29.639645]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [-0.      ]
 [30.071278]]
```
